### DISCUSSION TITLE
Preparing Alpha Release and getting crashes with web interface

### DISCUSSION THREAD

**Discussion by albar965** [albar965/littlenavmap#848] [General]
Hi @u-an-i 

I'm getting a reproducible crash on Windows  using the web interface in the current development branch (master).
The exception is thrown in `MapQuery::getIls(GeoDataLatLonBox rect, const MapLayer *mapLayer, bool lazy, bool& overflow)`.
![grafik](https://user-images.githubusercontent.com/15254956/153031683-feffb975-b7f9-4a9f-b153-f3c57d7a0e1c.png)
On Linux I get random crashes using the web interface in `MapPaintWidget::getPixmap()`.

The normal GUI uses the same query method without issues. Is this method called from another thread by any chance?

Alex

**Comment by KOKAProduktion** [comment]
Hi @albar965, hi @u-an-i,
maybe this issue has been introduced by #772 and #776 

https://github.com/albar965/littlenavmap/pull/826#discussion_r801860346

Cheers!

**Comment by u-an-i** [comment]
Hi @albar965 ,

the web interface had only been "streamlined" by me, no threaded additions, the only multi-threading occurs in the exe gui random flight planner. I don't remember calling any new methods, in particular not `getIls`.

@KOKAProduktion 's comment is about a not-merged code as far as I see.

Does the crash occur as soon as you load the web interface?

**Comment by u-an-i** [comment]
The SQL query appears fine and you appear to bind 4 values while having 4 placeholders, I don't see an issue there either. Is the datatype perhaps not matching somehow?

**Comment by albar965** [comment]
> I don't remember calling any new methods, in particular not getIls
This happens deep in the MarbleWidget guts when painting. The number of placeholders in the query is ok. That's why the error message is really confusing. This is crash one which happens on Windows.

> Does the crash occur as soon as you load the web interface?
Yup.

The getPixmap() thing (number two) can be reproduced on Linux but it happens only randomly.

**Reply by u-an-i** [reply]
hmm, i have no idea, but i'll look into the code.

This is the moment I must actually setup my development environment again :| . I haven't done so far on my new machine i built towards end of last year.

@albar965 Which Qt version are you using now?

**Comment by albar965** [comment]
> Is the datatype perhaps not matching somehow?
I doubt it. The same query is used in the GUI and it's working fine in Linux.

Just trying to reproduce it on Linux where I have the debugger but nothing.

**Comment by albar965** [comment]
> Which Qt version are you using now?

5.15.2

**Reply by u-an-i** [reply]
Qt open source installer offers pre built 5.15.2 components only for MingW 8.1.0. Previously we used MingW 7.3.0
I selected "sources" only for 5.15.2 but Qt Creator doesn't find any Qt with those.
I now let install MingW 8.1.0. I report back when i get Qt working again.

**Reply by u-an-i** [reply]
Qt Creator now finds Qt.

I receive an error when building from master branch:

```
C:\Users\uani\OneDrive\Dokumente\GitHub\littlenavmap\src\mapgui\mappaintwidget.cpp:322: Fehler: 'setKeys' was not declared in this scope
..\littlenavmap\src\mapgui\mappaintwidget.cpp: In member function 'virtual void MapPaintWidget::optionsChanged()':
..\littlenavmap\src\mapgui\mappaintwidget.cpp:322:3: error: 'setKeys' was not declared in this scope
   setKeys(mapThemeHandler->getMapThemeKeysHash());
   ^~~~~~~
..\littlenavmap\src\mapgui\mappaintwidget.cpp:322:3: note: suggested alternative: 'setTheme'
   setKeys(mapThemeHandler->getMapThemeKeysHash());
   ^~~~~~~
   setTheme
```

Which branch do I need to build from? @albar965

**Comment by albar965** [comment]
You need to rebuild Marble or use this one: https://magentacloud.de/s/Lfk2LAm7ie7Weia

What is use is `if defined PATH_SHARED ( echo %PATH_SHARED% ) else ( set PATH_SHARED=C:\Qt\5.15.2\mingw81_32\bin;C:\Qt\Tools\mingw810_32\bin)`.

**Reply by u-an-i** [reply]
i am not becoming wiser through this reply :) `setKeys` is related to Marble? Anyways, I started looking into the code. Unfortunately I don't understand the `getIls` method, but i see this warning about the data which can get detached (ie. leaked I think): 
![image](https://user-images.githubusercontent.com/84718885/153059561-76f85ddd-2053-447b-ade2-2a8e5ebd5f96.png)
Could this be the cause? A recommended workaround is to use "Fix it by marking the container const, or, since Qt 5.7, use qAsConst():" according to https://github.com/KDE/clazy/blob/1.10/docs/checks/README-range-loop-detach.md

@albar965

**Reply by albar965** [reply]
Yup. setKey is a new method in the Marble widget used by LNM to allow the usage or API keys or tokens for maps.
The detached containers are only a performance issue since a new deep copy is created. I get rid of these wherever possible. 

I'll check the query in the meantime. Maybe an issue with old databases?

**Comment by albar965** [comment]
What I get on Linux now:
```
[08 20:35:44.362 DEBUG] httprequest.cpp:42: HttpRequest: from ::ffff:192.168.2.14: GET /plugins HTTP/1.1
[08 20:35:47.353 DEBUG] httprequest.cpp:42: HttpRequest: from ::ffff:192.168.2.14: GET /refresh?session&maprefresh=0 HTTP/1.1
[08 20:35:47.361 DEBUG] httprequest.cpp:42: HttpRequest: from ::ffff:192.168.2.14: GET /mapimage?format=jpg&quality=70&width=1800&height=878&session&reload=0.43638072606285894 HTTP/1.1
[08 20:35:47.362 WARN ] webtools.cpp:93: float Parameter::asFloat() Wrong float value for "distance" : ""
[08 20:35:47.362 DEBUG] mappaintwidget.cpp:239: setting map theme to index 8 "earth/opentopomap/opentopomap.dgml"
[08 20:35:47.403 CRIT ] application.cpp:80: receiver ""
[08 20:35:47.403 CRIT ] application.cpp:81: event 12
[08 20:35:47.403 CRIT ] application.cpp:118: Caught exception in file ../atools/src/gui/application.cpp line 83 what Parameter count mismatch. SqlQuery::exec(): Error executing query. Query is "select ils_id, ident, name, region, mag_var, loc_heading, has_backcourse, loc_runway_end_id, loc_airport_ident, loc_runway_name, gs_pitch, frequency, range, dme_range, loc_width, end1_lonx, end1_laty, end_mid_lonx, end_mid_laty, end2_lonx, end2_laty, altitude, lonx, laty from ils where lonx between :leftx and :rightx and laty between :bottomy and :topy limit 8000". Bound values are [":bottomy"="49.26791770677447",":leftx"="-5.630273737011235",":rightx"="2.822664736596295",":topy"="52.28145116397077"].
20:35:47: The program has unexpectedly finished.
```

Maybe my mess. Have to see. I usually try to keep LNM compatible with old scenery library databases. Maybe this is an issue here.

**Comment by albar965** [comment]
The plot thickens. Seems to happen only after switching databases.

**Comment by albar965** [comment]
Never mind. I think I just got it fixed. The databases were not updated in the MapPaintWidget copy when switching between simulators. I added the usual `preDatabaseLoad()` and `postDatabaseLoad()` to fix this. I'll call it a day now and will commit tomorrow.
Sorry for the fuzz.

Alex

**Reply by u-an-i** [reply]
i'm reliefed you appear to have found the culprit. 👍

**Reply by albar965** [reply]
5c56dda6e7c4b9e267f9ec0e15a1aee5b0243004 got rid of the crashes.
And the maps requiring API keys or tokens also work now in the web interface.
Alex
